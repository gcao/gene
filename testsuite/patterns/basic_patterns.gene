#!/usr/bin/env gene run
# TEST: Pattern matching

# Simple value matching
(fn describe_number [n]
  (match n
    0 "zero"
    1 "one"
    2 "two"
    _ "other"))

(print "match 0: " (describe_number 0))
(print "match 1: " (describe_number 1))
(print "match 5: " (describe_number 5))

# Array pattern matching
(fn sum_pair [arr]
  (match arr
    [a b] (+ a b)
    [a] a
    [] 0
    _ "not a valid pair"))

(print "sum [3 4]: " (sum_pair [3 4]))
(print "sum [5]: " (sum_pair [5]))
(print "sum []: " (sum_pair []))

# Map pattern matching
(fn get_name [person]
  (match person
    {"name" n} n
    _ "Anonymous"))

(print "name from map: " (get_name {"name" "Alice" "age" 30}))
(print "name missing: " (get_name {"age" 25}))

# Guard clauses
(fn classify [x]
  (match x
    n when (and (int? n) (> n 0)) "positive integer"
    n when (and (int? n) (< n 0)) "negative integer"
    0 "zero"
    s when (string? s) "string"
    _ "other"))

(print "classify 5: " (classify 5))
(print "classify -3: " (classify -3))
(print "classify 'hello': " (classify "hello"))