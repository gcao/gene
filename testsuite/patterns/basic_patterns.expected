match 0: zero
match 1: one
match 5: other
sum [3 4]: 7
sum [5]: 5
sum []: 0
name from map: Alice
name missing: Anonymous
classify 5: positive integer
classify -3: negative integer
classify 'hello': string